The SeacrhPanelSecondary has been changed with react-router

min-in-zon opened the pull request:
### Github issue

- Issue URL: #65 

### Description

- Changed the functionality of search panels from using context to using react-router.

### Checklist

- [x] SearchPanelPrimary has been changed
- [x] SearchPanelSecondary has been changed
- [x] The transition from the main page to the page with cities has been configured, while maintaining the search query.
- [x] useSearchQuery is created. This hook is designed to manage the state of the search string (query) and synchronize it with the URL.
- [x] Removed unused contexts
- [x] Changed placeholder text to less contrast to improve user experience


### What is the current behavior?

### What is the new behavior?

![image](https://github.com/user-attachments/assets/ef9bd738-2264-4c3a-be6c-957e49a4575f)
![image](https://github.com/user-attachments/assets/41f4fba9-1904-4d34-a242-f52431aa160d)

adadychanka commented:
You need to handle these two links as well, so they include query search param

![image](https://github.com/user-attachments/assets/1e05f2d4-323b-4175-8ce6-b9035414c156)

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+import { useState, useEffect, useCallback } from 'react';
+import { useNavigate, useSearchParams } from 'react-router-dom';
+
+export const useSearchQuery = () => {
```
The first thing I don't like is the name because it is too generic. In fact, the hook is used only for making search query for cities search so let's name it properly.

`useCitySearchQuery` should be better

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+  const [searchParams] = useSearchParams();
+
+  useEffect(() => {
+    const queryParam = searchParams.get('query') || '';
```
Make the `query` string a constant so you can reuse it somewhere else.

adadychanka replied:
and let's name it more precise, `citySearchQuery` or something

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+import { useNavigate, useSearchParams } from 'react-router-dom';
+
+export const useSearchQuery = () => {
+  const [query, setQuery] = useState('');
```
There is unnecessary complexity in having a local state for the search param as well. You don't need it actually and you can just rely on the value from the search param

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+    setQuery(queryParam);
+  }, [searchParams]);
+
+  const updateQuery = useCallback(
```
to update search param you should use the setSearchParams function returned from `useSearchParams()` hook.

https://reactrouter.com/6.28.1/hooks/use-search-params

adadychanka commented:
Check that article about SRP

https://solidbook.vercel.app/srp

adadychanka commented on src/components/common/SearchPanels/SearcPanelSecondary.jsx:
You have a typo in file name

adadychanka commented on src/components/common/SearchPanels/SearcPanelSecondary.jsx:
```diff
@@ -16,11 +16,13 @@ const SearchPanelSecondary = () => {
-        placeholder="India | Deli"
+        placeholder="India | Delhi"
         aria-label="Search for guides by city or country name"
+        value={query}
```
You can try using the combination `defaultValue` + taking value from input ref

It might help to throw out the logic storing the value because we actually need it only when the user clicks submit button

adadychanka commented on src/hooks/useAllCities.js:
```diff
@@ -1,33 +1,35 @@
 
-const useAllCities = ({ pageSize, sortOrder, searchQuery }) => {
+const useAllCities = ({ pageSize, sortOrder }) => {
+  const [searchParams] = useSearchParams();
```
Let's get back to old approach when you pass searchQuery as an argument

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+      const trimmedQuery = newQuery.trim();
+      if (trimmedQuery) {
+        searchParams.set('query', trimmedQuery);
+        navigate(`${route}?${searchParams.toString()}`);
```
just in case, navigate can accept object, so you don't need to construct the whole URL by yourself

`navigate({ pathname: '/SOME_PATH_IN_APP', search: '?somekey=somevalue' })`

if you for some reason want to create search params by yourself, you can use https://reactrouter.com/6.28.1/utils/create-search-params

adadychanka commented on src/hooks/useSearchQuery.jsx:
```diff
@@ -0,0 +1,29 @@
+        navigate(`${route}?${searchParams.toString()}`);
+      } else {
+        searchParams.delete('query');
+        navigate(route || '?');
```
it might be better to do nothing instead of sending a user to /?

adadychanka commented on src/pages/CitiesPage/CitiesSection.jsx:
```diff
@@ -1,16 +1,14 @@
 const CitiesSection = () => {
-  const { searchQuery } = useCitySearch();
-  const { citiesData, isDataLoaded } = useCitiesData();
+  const [searchParams] = useSearchParams();
```
you can reuse the hook you created

min-in-zon commented:
1) Renamed files and variables in useCitySearchQuery
2) Revised the code and applied development principles such as SOLID and SRP, KISS and etc.
3) useCitySearchQuery was split into two different hooks, as they have different functionality (+ useNavigateToRoute)
4) Removed storage in local state in the hook.
5) Removed memo in SearchPanelPrimary and Secondary, as there is no reason to use it in the component (props, etc.)
6) useState was added directly to the SearchPanelPrimary component to correctly configure the value update in the form input. Without it, it is very problematic to configure the value update in the form and send a request to the server only when the enter button is pressed.
7) Added icons from Lucid.
8) Returned the searchQuery argument to the useAllCities props.
9) Created constant CITY_SEARCH_QUERY for query.
10) Added transitions by links (to city and to country) inside the city page.
Transition by clicking on the country is not final. Now it uses useNavigateToRoute and goes to the page with cities and substitutes query = countryName, in the future a hook will be created, useCountryID, to use the backend functionality.

adadychanka commented on src/components/common/SearchPanels/SearchPanelPrimary.jsx:
```diff
@@ -1,25 +1,50 @@
+
+    updateCitySearchQuery(queryInput);
+
+    navigateToRoute('/cities', queryInput);
```
`ROUTES.cities` can be used here

adadychanka commented on src/components/common/SearchPanels/SearchPanelPrimary.jsx:
```diff
@@ -1,25 +1,50 @@
-import SearchIcon from '../../../assets/img/searchIcon.svg';
```
If it isn't used anywhere else, the asset can be deleted

adadychanka commented on src/pages/CityPage/CityPageHeader.jsx:
```diff
@@ -14,18 +16,26 @@ const CityPageHeader = () => {
           <h2 className="inline-block pr-4 font-third text-4xl">{cityName}</h2>
 
-          <Link
+          <a
```
a + href = bad
link + to string = better
link + to object = the best

adadychanka replied:
```tsx
          <Link
            to={{
              pathname: ROUTES.cities,
              search: createSearchParams({ query: [countryName] }).toString(),
            }}
            className="inline-block font-primary text-2xl underline transition duration-300 ease-in-out hover:text-orange-color"
          >
            {countryName}
          </Link>
```

adadychanka replied:
Ideal

```tsx
          <Link
            to={{
              pathname: ROUTES.cities,
              search: createSearchParams({
                [CITY_SEARCH_QUERY]: [countryName],
              }).toString(),
            }}
            className="inline-block font-primary text-2xl underline transition duration-300 ease-in-out hover:text-orange-color"
          >
            {countryName}
          </Link>
```

adadychanka commented on src/pages/CityPage/CityPageHeader.jsx:
```diff
@@ -14,18 +16,26 @@ const CityPageHeader = () => {
   const countryName = city?.country?.name || '';
   const cityDescription = city?.description;
 
+  const handleCountryClick = event => {
```
don't need it if you have correct link

adadychanka commented on src/hooks/useNavigateToRoute.jsx:
I see this hook as an unnecessary abstraction. It doesn't bring new functionality into the navigate function or combine it with something existing. Basically, you just changed the API of `useNavigate`.

The given name is confusing. It looks abstract but it is not. It serves to navigation to cities pages with search query.

Here are 2 pathes to resolve this:
1) You remove `useNavigateToRoute` and use `useNavigate` from router
2) You modify and rename your hook so the purpose of the hook is reflected in the name

adadychanka replied:
If you want to stick #2, you can:

Omit the first param in `navigateToRoute` and hardcode `pathname` to `ROUTES.citites` because you actually need this search query key for only one page.

The hook can be renamed to `useNavigateToSearchCities` or something

adadychanka commented on src/hooks/useCitySearchQuery.jsx:
```diff
@@ -0,0 +1,30 @@
+import { useCallback, useMemo } from 'react';
+import { useSearchParams } from 'react-router-dom';
+
+const CITY_SEARCH_QUERY = 'query';
```
The idea was to have it as constant to reuse it.

The usage across the project should be like this 
**before**
```js
search: `query=${cityName}`
```
after
```
search: createSearchParams({
                [CITY_SEARCH_QUERY]: [countryName],
              }).toString()
```
or at least
```
search: `${CITY_SEARCH_QUERY}=${cityName}`
```
